## Supervised Learning

When inputs and their expected outputs are given to ML model during the training.

- Expected output = Label = Target

To apply supervised learning, you should have labelled data.

### 0. Import "three musketeers'" of the ML world!

In [9]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")

### 1. Load the data

In [1]:
from sklearn.datasets import load_breast_cancer

In [2]:
data = load_breast_cancer()

In [10]:
df = pd.DataFrame(data.data, columns=data.feature_names)

In [12]:
df['label'] = data.target

In [13]:
df.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  label  
0          0.4601                  0.11890      0  
1          0.2750                  0.08902      0  
2          0.3613                  0.08758      0  
3          0.6638                  0.17300      0  
4          0.2364                  0.07678      0  

[5 rows x 31 columns]

### 2. Explore the data

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 31 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   mean radius              569 non-null    float64
 1   mean texture             569 non-null    float64
 2   mean perimeter           569 non-null    float64
 3   mean area                569 non-null    float64
 4   mean smoothness          569 non-null    float64
 5   mean compactness         569 non-null    float64
 6   mean concavity           569 non-null    float64
 7   mean concave points      569 non-null    float64
 8   mean symmetry            569 non-null    float64
 9   mean fractal dimension   569 non-null    float64
 10  radius error             569 non-null    float64
 11  texture error            569 non-null    float64
 12  perimeter error          569 non-null    float64
 13  area error               569 non-null    float64
 14  smoothness error         5

In [17]:
df.describe().T

count        mean         std         min  \
mean radius              569.0   14.127292    3.524049    6.981000   
mean texture             569.0   19.289649    4.301036    9.710000   
mean perimeter           569.0   91.969033   24.298981   43.790000   
mean area                569.0  654.889104  351.914129  143.500000   
mean smoothness          569.0    0.096360    0.014064    0.052630   
mean compactness         569.0    0.104341    0.052813    0.019380   
mean concavity           569.0    0.088799    0.079720    0.000000   
mean concave points      569.0    0.048919    0.038803    0.000000   
mean symmetry            569.0    0.181162    0.027414    0.106000   
mean fractal dimension   569.0    0.062798    0.007060    0.049960   
radius error             569.0    0.405172    0.277313    0.111500   
texture error            569.0    1.216853    0.551648    0.360200   
perimeter error          569.0    2.866059    2.021855    0.757000   
area error               569.0   40.337079   45.491006    6.802000   
smoothness error         569.0    0.007041    0.003003    0.001713   
compactness error        569.0    0.025478    0.017908    0.002252   
concavity error          569.0    0.031894    0.030186    0.000000   
concave points error     569.0    0.011796    0.006170    0.000000   
symmetry error           569.0    0.020542    0.008266    0.007882   
fractal dimension error  569.0    0.003795    0.002646    0.000895   
worst radius             569.0   16.269190    4.833242    7.930000   
worst texture            569.0   25.677223    6.146258   12.020000   
worst perimeter          569.0  107.261213   33.602542   50.410000   
worst area               569.0  880.583128  569.356993  185.200000   
worst smoothness         569.0    0.132369    0.022832    0.071170   
worst compactness        569.0    0.254265    0.157336    0.027290   
worst concavity          569.0    0.272188    0.208624    0.000000   
worst concave points     569.0    0.114606    0.065732    0.000000   
worst symmetry           569.0    0.290076    0.061867    0.156500   
worst fractal dimension  569.0    0.083946    0.018061    0.055040   
label                    569.0    0.627417    0.483918    0.000000   

                                25%         50%          75%         max  
mean radius               11.700000   13.370000    15.780000    28.11000  
mean texture              16.170000   18.840000    21.800000    39.28000  
mean perimeter            75.170000   86.240000   104.100000   188.50000  
mean area                420.300000  551.100000   782.700000  2501.00000  
mean smoothness            0.086370    0.095870     0.105300     0.16340  
mean compactness           0.064920    0.092630     0.130400     0.34540  
mean concavity             0.029560    0.061540     0.130700     0.42680  
mean concave points        0.020310    0.033500     0.074000     0.20120  
mean symmetry              0.161900    0.179200     0.195700     0.30400  
mean fractal dimension     0.057700    0.061540     0.066120     0.09744  
radius error               0.232400    0.324200     0.478900     2.87300  
texture error              0.833900    1.108000     1.474000     4.88500  
perimeter error            1.606000    2.287000     3.357000    21.98000  
area error                17.850000   24.530000    45.190000   542.20000  
smoothness error           0.005169    0.006380     0.008146     0.03113  
compactness error          0.013080    0.020450     0.032450     0.13540  
concavity error            0.015090    0.025890     0.042050     0.39600  
concave points error       0.007638    0.010930     0.014710     0.05279  
symmetry error             0.015160    0.018730     0.023480     0.07895  
fractal dimension error    0.002248    0.003187     0.004558     0.02984  
worst radius              13.010000   14.970000    18.790000    36.04000  
worst texture             21.080000   25.410000    29.720000    49.54000  
worst perimeter           84.110000   97.660000   125.40000

- No missing value
- All numerical
- 30 features & 569 rows
- Not normalized

In [20]:
df['label'].value_counts(normalize=True)

label
1    0.627417
0    0.372583
Name: proportion, dtype: float64

- Binary classification
- Class distribution 63-37 => Not balanced and not severe imbalanced.

In [22]:
# missing value check
df.isnull().sum()

mean radius                0
mean texture               0
mean perimeter             0
mean area                  0
mean smoothness            0
mean compactness           0
mean concavity             0
mean concave points        0
mean symmetry              0
mean fractal dimension     0
radius error               0
texture error              0
perimeter error            0
area error                 0
smoothness error           0
compactness error          0
concavity error            0
concave points error       0
symmetry error             0
fractal dimension error    0
worst radius               0
worst texture              0
worst perimeter            0
worst area                 0
worst smoothness           0
worst compactness          0
worst concavity            0
worst concave points       0
worst symmetry             0
worst fractal dimension    0
label                      0
dtype: int64

- no missing value

In [24]:
# duplicate check
df.duplicated().sum()

0

- no duplicate

### 4. Split the data

In [18]:
from sklearn.model_selection import train_test_split

In [25]:
X = df.drop('label', axis=1)

In [26]:
y = df['label']

In [27]:
X.shape

(569, 30)

In [28]:
y.shape

(569,)

In [48]:
X_train, X_test, y_train, y_test = train_test_split(X, y, stratify=y, test_size=0.2, random_state=42)

In [33]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(455, 30)
(114, 30)
(455,)
(114,)


### 5. Feature improvement

In [42]:
# normalization
from sklearn.preprocessing import StandardScaler, RobustScaler, MinMaxScaler

In [55]:
scaler = StandardScaler()

In [56]:
X_train_scaled = scaler.fit_transform(X_train)

In [57]:
X_test_scaled = scaler.transform(X_test)

### 6. Build Logistic Regression Model

In [58]:
from sklearn.linear_model import LogisticRegression

In [59]:
log_model = LogisticRegression()

In [60]:
log_model.fit(X_train_scaled, y_train)

LogisticRegression()

In [62]:
y_pred = log_model.predict(X_test_scaled)

### 7. Evaluate the model

In [63]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix

In [65]:
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))

Accuracy: 0.9824561403508771
Precision: 0.9861111111111112
Recall: 0.9861111111111112


In [66]:
confusion_matrix(y_test, y_pred)

array([[41,  1],
       [ 1, 71]])

In [67]:
from sklearn.metrics import classification_report

In [69]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.98      0.98      0.98        42
           1       0.99      0.99      0.99        72

    accuracy                           0.98       114
   macro avg       0.98      0.98      0.98       114
weighted avg       0.98      0.98      0.98       114



In [93]:
# roc auc
from sklearn.metrics import roc_auc_score

In [94]:
y_probs = log_model.predict_proba(X_test_scaled)[:, 1]

In [96]:
print("ROC AUC score:", roc_auc_score(y_test, y_probs))

ROC AUC score: 0.9953703703703703


In [88]:
# cross validation
from sklearn.model_selection import cross_validate

In [89]:
scoring = ['accuracy', 'precision', 'recall']
cv_results = cross_validate(model, X, y, cv=5, scoring=scoring)

In [90]:
acc_scores = cv_results['test_accuracy']
precision_scores = cv_results['test_precision']
recall_scores = cv_results['test_recall']

In [91]:
print(acc_scores.mean())
print(precision_scores.mean())
print(recall_scores.mean())

0.9525694767893185
0.951826432425003
0.9748043818466353


### 8. Compare the scalers

In [35]:
# define scalers
scalers = [('StandardScaler', StandardScaler()),
           ('MinMaxScaler', MinMaxScaler()),
           ('RobustScaler', RobustScaler())]

In [36]:
from sklearn.pipeline import Pipeline

In [37]:
from sklearn.model_selection import cross_val_score

In [41]:
for scaler in scalers:
    pipeline = Pipeline([scaler,
                         ('logistic_regression', LogisticRegression())])
    scores = cross_val_score(pipeline, X, y, cv=5)
    print(f"{scaler[0]}: Mean Accuracy: {scores.mean():.4f}, Std Dev: {scores.std():.4f}")

StandardScaler: Mean Accuracy: 0.9807, Std Dev: 0.0065
MinMaxScaler: Mean Accuracy: 0.9613, Std Dev: 0.0042
RobustScaler: Mean Accuracy: 0.9789, Std Dev: 0.0089


### 9. Hyper-parameter Optimization

In [70]:
from sklearn.model_selection import GridSearchCV

In [78]:
# hyperparameters
param_grid = {
    'penalty': ['l1', 'l2'],
    'C': [0.001, 0.01, 0.1, 1, 10, 100],
    'solver': ['liblinear', 'lbfgs', 'newton-cholesky','saga']
}

In [79]:
model = LogisticRegression(max_iter=1000)
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5, scoring='accuracy')


In [80]:
grid_search.fit(X_train_scaled, y_train)

GridSearchCV(cv=5, estimator=LogisticRegression(max_iter=1000),
             param_grid={'C': [0.001, 0.01, 0.1, 1, 10, 100],
                         'penalty': ['l1', 'l2'],
                         'solver': ['liblinear', 'lbfgs', 'newton-cholesky',
                                    'saga']},
             scoring='accuracy')

In [81]:
best_params = grid_search.best_params_
print("Best Hyperparameters:", best_params)

Best Hyperparameters: {'C': 0.1, 'penalty': 'l2', 'solver': 'lbfgs'}


In [82]:
best_model = grid_search.best_estimator_

In [83]:
y_pred = best_model.predict(X_test_scaled)

In [84]:
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))

Accuracy: 0.9736842105263158
Precision: 0.9726027397260274
Recall: 0.9861111111111112


In [85]:
confusion_matrix(y_test, y_pred)

array([[40,  2],
       [ 1, 71]])

### 10. Save the model

In [97]:
import joblib

In [98]:
joblib.dump(model, 'logistic_regression_model.pkl')

['logistic_regression_model.pkl']